core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java

```diff
@@ -0,0 +1,330 @@
+        var isChanged = false;
+        var entity = shadowVariable.entity();
+        var shadowVariableReference = shadowVariable.variableReference();
+        var oldValue = shadowVariableReference.memberAccessor().executeGetter(entity);
+
+        if (isLooped) {
+            // null might be a valid value, and thus it could be the case
+            // that is was not looped and null, then turned to looped and null,
+            // which is still considered a change.
+            isChanged = true;
+            affectedEntities.add(new AffectedEntity(entity, shadowVariableReference));
+            if (oldValue != null) {
+                changedVariableNotifier.beforeVariableChanged().accept(
+                        (VariableDescriptor<Solution_>) shadowVariableReference.variableDescriptor(), entity);
+                shadowVariableReference.memberAccessor().executeSetter(entity, null);
+                changedVariableNotifier.afterVariableChanged().accept(
+                        (VariableDescriptor<Solution_>) shadowVariableReference.variableDescriptor(), entity);
+            }
+        } else {
+            var newValue = shadowVariableReference.calculator().apply(entity);
+
+            if (!Objects.equals(oldValue, newValue)) {
+                affectedEntities.add(new AffectedEntity(entity, shadowVariableReference));
+                changedVariableNotifier.beforeVariableChanged().accept(
+                        (VariableDescriptor<Solution_>) shadowVariableReference.variableDescriptor(), entity);
+                shadowVariableReference.memberAccessor().executeSetter(entity, newValue);
+                changedVariableNotifier.afterVariableChanged().accept(
+                        (VariableDescriptor<Solution_>) shadowVariableReference.variableDescriptor(), entity);
+                isChanged = true;
+            }
+        }
+        return isChanged;
+    }
+
+    record AffectedShadowVariable(int nodeId, int topologicalIndex) implements Comparable<AffectedShadowVariable> {
+        @Override
+        public int compareTo(AffectedShadowVariable heapItem) {
+            return topologicalIndex - heapItem.topologicalIndex;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (o instanceof AffectedShadowVariable other) {
+                return nodeId == other.nodeId;
+            }
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return nodeId;
+        }
+    }
+
+    private PriorityQueue<AffectedShadowVariable> createInitialChangeQueue() {
+        var heap = new PriorityQueue<AffectedShadowVariable>(instanceList.size());
+        for (var i = changed.nextSetBit(0); i >= 0; i = changed.nextSetBit(i + 1)) {
+            var topologicalOrder = graph.getTopologicalOrder(i);
+            heap.add(new AffectedShadowVariable(i, topologicalOrder));
+            if (i == Integer.MAX_VALUE) {
+                break; // or (i+1) would overflow
```

Comment:
Shouldn't we rather throw an exception here?
Seems like this is a failure state, maybe even impossible state.

Reply:
Overly cautious `BitSet` iteration. How `BitSet` iteration works is by querying for the first set bit after or at an index. Then to get the next set bit, you do `nextSetBit(currentBit + 1)`. `currentBit + 1` will overflow if BitSet has the INT_MAX bit set. For our cases, this is impossible, since the JVM limit array sizes to slightly less than INT_MAX.

Reply:
I think you've just given a reason for why this should have been treated as a failure state. :-) At the very least, there should have been a comment explaining what you just described.

Try to think of it like this: if I have a question because I don't understand something, it's probably a piece of code that deserves an explanatory comment. :-) There already is a comment, but in this case, the additional context you provided is useful.